### PR TITLE
Move platforms software configuration from Zephyr to SOF

### PR DESCRIPTION
**Pull request by aborisovich**
Move software configuration from Zephyr samples directory to SOF platforms directory using Zephyr [overlay](https://docs.zephyrproject.org/latest/application/index.html#important-build-system-variables) mechanism.

Zephyr .conf files in /zephyr/samples/subsys/audio/sof will be deleted after this PR is integrated.
Temporary configuration duplication does not break anything.
This PR only moves configuration from Zephyr without any further refactorings, this will be done later.
Existing IPC4 overlay for TGL is merged with default overlay.conf achieving expected result.

Existing thread-race with [PR5503: xtensa-build-all: add a separate target for JSL](https://github.com/thesofproject/sof/pull/5503): added only overlay.conf for JSL (for now).

Signed-of-by: Andrey Borisovich <andrey.borisovich@intel.com>

### CODE REVIEW

File: platforms/apl/overlay.conf
Comment (dbaluta) [inline platforms/apl/overlay.conf:2]: Can you explain a little bit what is an overlay? Is just a fragment of config file that gets applied over default config?

File: scripts/xtensa-build-zephyr.py
Comment (dbaluta) [inline scripts/xtensa-build-zephyr.py:35]: This means that we won't support xtensa-build-zephyr.sh script anymore?

Cc: @marc-hb

File: platforms/apl/overlay.conf
Comment (aborisovich) [reply platforms/apl/overlay.conf:2]: Exactly. It just applies one by one over existing Kconfig entries.
Let me just paste part of the build output logs:
`Parsing /home/aborisov/workspace/software_configuration_pub/zephyrproject/zephyr/Kconfig
Loaded configuration '/home/aborisov/workspace/software_configuration_pub/zephyrproject/zephyr/boards/xtensa/intel_adsp_cavs25/intel_adsp_cavs25_defconfig'
Merged configuration '/home/aborisov/workspace/software_configuration_pub/zephyrproject/zephyr/samples/subsys/audio/sof/prj.conf'
Merged configuration '/home/aborisov/workspace/software_configuration_pub/platforms/tgl/overlay.conf'
Configuration saved to '/home/aborisov/workspace/software_configuration_pub/zephyrproject/build/zephyr/.config'
Kconfig header saved to '/home/aborisov/workspace/software_configuration_pub/zephyrproject/build/zephyr/include/generated/autoconf.h'`

This is TGL build example. So it basically:
1. Gets zephyr board defconfig values
2. Then applies prj.conf values on top of it
3. Then applies our new overlay.conf on top of it
<etc... may be further overlay files provided that will just overwrite values that are repeated>
Final "summary" of Kconfig settings is written to autoconf.h header file.

Comment (dbaluta) [reply platforms/apl/overlay.conf:2]: @aborisovich this is very good information. If you can add this  please in the commit message would be great!

**Comment by lgirdwood** [comment]
@marc-hb @kv2019i @keqiaozhang fyi in case config updates needed for CI or testing scripts.

File: scripts/xtensa-build-zephyr.py
Comment (lgirdwood) [reply scripts/xtensa-build-zephyr.py:35]: @dbaluta plan is to have one script and the python version is easier to maintain and develop (supports Windows too). Are you using the bash script internally or with yocto - if so we need to create a transition plan.

File: platforms/apl/overlay.conf
Comment (aborisovich) [reply platforms/apl/overlay.conf:1]: Sure I will reword in a moment.

> Another, thing is that you are creating a new directory src/platforms right? We already have src/platform can you somehow use this one instead?

This is totally not related to the sources I do not think this is good idea.

File: scripts/xtensa-build-zephyr.py
Comment (dbaluta) [reply scripts/xtensa-build-zephyr.py:35]: @lgirdwood thanks for the info. We are fine with using the python script with next SOF release. Should be v2.1 right?

BTW, any schedule when that will be released?

File: platforms/apl/overlay.conf
Comment (dbaluta) [reply platforms/apl/overlay.conf:1]: @aborisovich @lgirdwood yes, but having two directories named `src/platform` and `src/platforms` is very confusing. We need to find a better location for this conf files.

Comment (aborisovich) [reply platforms/apl/overlay.conf:1]: This is not src/platforms.
It is just SOF_TOP/platforms.

Comment (dbaluta) [reply platforms/apl/overlay.conf:1]: @aborisovich oh, ok! It is fine then.

File: platforms/apl/overlay.conf
Comment (marc-hb) [reply platforms/apl/overlay.conf:2]: overlays are documented in https://docs.zephyrproject.org/latest/application/index.html

File: scripts/xtensa-build-zephyr.py
Comment (marc-hb) [reply scripts/xtensa-build-zephyr.py:35]: SOF stable-v2.1 has been branched, it's too late to deprecate xtensa-build-zephyr.sh in 2.1

xtensa-build-zephyr.sh should absolutely be deprecated for 2.2, much earlier in fact. It will take a bit of time to make sure  all CIs have switched but letting xtensa-build-zephyr.sh "bitrot" already  now is a good way to put pressure on CI.

File: scripts/xtensa-build-zephyr.py
Comment (marc-hb) [inline scripts/xtensa-build-zephyr.py:362]: I don't think it's possible to have an entirely empty Kconfig file, at the very least we need `CONFIG_ICELAKE=y` or equivalent.

So can you make `overlay.conf` the default filename and avoid the repetition? Something like this:
```python
overlay_filename = platform_dict.get("CONFIG_OVERLAY", default="overlay.conf")
overlays.append(str(pathlib.Path(SOF_TOP, "platforms", platform, overlay_filename))
```

**Review by marc-hb — COMMENTED** [review]
Thanks, the concept looks great. Some comments on the implementation details.

I'm not a fan of the poorly defined `platforms/` name (see #3491), how about a more basic `overlays/apl/`?

Who knows, maybe we will have device tree or some other overlays in the future, could you check https://docs.zephyrproject.org/latest/application/index.html and the existing Zephyr source and find something more specific than `overlay.conf`? Something with `kconfig` in the name maybe?

Speaking of naming conventions, could someone from the core Zephyr team take a look at this? @andyross , @dcpleung , @MaureenHelm , other? Thanks in advance!

Also, to test this it would be good to compare the `generated/.config` files with and without this PR.

**Comment by aborisovich** [comment]
Push update:
- Updated xtensa-build-zephyr.sh so it supports overlay files
- Reworded commit message according to @dbaluta suggestions.

File: scripts/xtensa-build-zephyr.py
Comment (aborisovich) [reply scripts/xtensa-build-zephyr.py:362]: Yes sure. I was wondering if it is possible to build without Kconfig entries at all just by what is given in board files but if it is not lets remove it.

**Comment by aborisovich** [comment]
Push update:

- Removed repetition of CONFIG_OVERLAY in platform dictionaries - replaced with default value in both .py and .sh scripts
- Fixed previous .sh script errors

**Comment by aborisovich** [comment]
@marc-hb 

> I'm not a fan of the poorly defined platforms/ name (see https://github.com/thesofproject/sof/issues/3491), how about a more basic overlays/apl/?

I've read briefly that improvement article. I do agree that there is a mess currently in the sense you described there but I do not agree that using platform abstraction is bad approach. Freely speaking if we would build computer mice's then the Zephyr board would be lets say "Logitech Master Family" and platform abstraction would be "Logitech Master MX3" and overlay.conf would contain information whether it uses micro USB or USB-C port to charge.
There were ideas (that you took part in :-)) to use "platforms" folder in future not only for Kconfig files but also for CMake files that would contain per-platform well known miscellaneous information - tt would be moved from xtensa-build-zephyr scripts to the build system. That is why @aiChaoSONG followed this approach in [xtensa-build-zephyr.py: add ipc4 build support for tgl](https://github.com/thesofproject/sof/commit/6f71808e3ee74a8782f1a65cf7154b7f04df6ea2).

> Who knows, maybe we will have device tree or some other overlays in the future, could you check https://docs.zephyrproject.org/latest/application/index.html and the existing Zephyr source and find something more specific than overlay.conf? Something with kconfig in the name maybe?

I've searched over Zephyr docs and overlays is the only way to go for us.
Only overlays allow to "append" additional Kconfig entries to the already existing set of them and we already have a _zephyrproject/zephyr/samples/subsys/audio/sof/prj.conf_ file that contains general "glue" settings for building Zephyr with SOF. So if you would like to add a Kconfig files somewhere you would have to repeat there content of that prj.conf file (I actually tried that) as it is not taken into account then. Shields also do not meet our needs as they are more of a different hardware version of something than a software configuration. "overlay.conf" is the default name they describe in the docs in examples. I also saw many applications using it this way (so you instantly see that this is an overlay not some other unknown conf file). I would stick with that.

Overlays are fine as we can concatanate them and the order of appearance resolves final value of the Kconfig entry.
So we can lets say build tgl with default overlay.conf and when we wish to build with overlay.conf+ipc4_overlay.conf that just alters some settings from the original set. Same can be done for TGL-H probably.

File: platforms/jsl/overlay.conf
Comment (kv2019i) [inline platforms/jsl/overlay.conf:4]: These are a bit problematic as these properties are not SW configuration, but HW properties that should be set on Zephyr side. 

But alas, these are the SOF side definitions and they are defined in the "samples/" dir in Zephyr, so I guess better to move them back to SOF now.

**Review by kv2019i — APPROVED** [review]
Minor spelling issues "Moved software" in git message body, but otherwise looks great, thanks @aborisovich .

@dbaluta So software configuration is basicly application specific Kconfig options, and specifically from Zephyr point of view. The split is very complicated at this point, as we have drivers in both SOF and Zephyr trees, so a setting lke CONFIG_INTEL_SSP does not make much sense (this is not an application property, but rather a hw property that in Zephyr should be described in DT). But e.g. CONFIG_COMP_SRC_TINY is a better example. This kind of setting does not belong to Zephyr as this is not a hardware property, but rather application (=SOF) property. This patch allows us to more cleanly separate descriing hw (that should be in Zephyr tree) and describing SOF application configuration (that should be in SOF tree).

**Comment by marc-hb** [comment]
> I've searched over Zephyr docs and I believe overlays is the only way to go for us.

@aborisovich I think we are talking across each other a bit. I was only commenting about the directory and file names, not suggesting any alternative to overlays at all (I'm not knowledgeable enough to suggest alternatives to overlay).

> There were ideas (that you took part in :-)) to use "platforms" folder in future not only for Kconfig files but also for CMake files that would contain per-platform well known miscellaneous information - it would be moved from xtensa-build-zephyr scripts to the build system. 

If you look at the toolchain parameters right now you can see that toolchain configuration is not per "platform": the same toolchain configuration covers many different "platforms". This new overlay system and re-organization is a great opportunity to drop the pseudo per "platform" classification and reduce copy/paste/diverge (which causes accidents like b9d971920e3032c)

> "overlay.conf" is the default name they describe in the docs in examples. I also saw many applications using it this way (so you instantly see that this is an overlay not some other unknown conf file). I would stick with that.


I think things like `overlays/tgl/defconfig` or `overlays/tgl_defconfig`, `overlays/tgl/app.conf` `overlays/xcc_2016.conf` would be more future proof and make more sense than the often very vague "platform" term. It would be even more obvious that they are overlays and they would drop the undefined and broken "platform" concept (#3491). If you have a collection of files like these then it's very clear that the filename is a meaningless placeholder:

```
.
├── adl
│   └── overlay.conf
├── apl
│   └── overlay.conf
└── tgl
    └── overlay.conf
```

I haven't look at the Zephyr docs much but maybe they fail or didn't want to make the examples specific.

**Comment by aborisovich** [comment]
> > I've searched over Zephyr docs and I believe overlays is the only way to go for us.
> 
> @aborisovich I think we are talking across each other a bit. I was only commenting about the directory and file names, not suggesting any alternative to overlays at all (I'm not knowledgeable enough to suggest alternatives to overlay).
> 
> > There were ideas (that you took part in :-)) to use "platforms" folder in future not only for Kconfig files but also for CMake files that would contain per-platform well known miscellaneous information - it would be moved from xtensa-build-zephyr scripts to the build system.
> 
> If you look at the toolchain parameters right now you can see that toolchain configuration is not per "platform": the same toolchain configuration covers many different "platforms". This new overlay system and re-organization is a great opportunity to drop the pseudo per "platform" classification and reduce copy/paste/diverge (which causes accidents like [b9d9719](https://github.com/thesofproject/sof/commit/b9d971920e3032c8dbbe87483349df1e5155da2c))
> 
> > "overlay.conf" is the default name they describe in the docs in examples. I also saw many applications using it this way (so you instantly see that this is an overlay not some other unknown conf file). I would stick with that.
> 
> I think things like `overlays/tgl/defconfig` or `overlays/tgl_defconfig`, `overlays/tgl/app.conf` `overlays/xcc_2016.conf` would be more future proof and make more sense than the often very vague "platform" term. It would be even more obvious that they are overlays and they would drop the undefined and broken "platform" concept (#3491). If you have a collection of files like these then it's very clear that the filename is a meaningless placeholder:
> 
> ```
> .
> ├── adl
> │   └── overlay.conf
> ├── apl
> │   └── overlay.conf
> └── tgl
>     └── overlay.conf
> ```
> 
> I haven't look at the Zephyr docs much but maybe they fail or didn't want to make the examples specific.

Marc the ambiguity of "platform" you talk about is changeset not an "issue" but a "plan".
I would describe what is happening here using following funny picture (if we replace "standard" with "platform")
image

The plan was to create some build.cmake file in platforms/<name>/ directory and place there all the things related to the given platform like rimage signing schema or what toolchain to use (what would reduce number of "platform" word occurrences in the project by putting those things into one place. But maybe you are right - I should not create things in this changeset that are "meant for future". If the directory has only overlays inside for now lets call it overlays! With further refactorings we can change that name as needed if other things will be put there.

**Comment by aborisovich** [comment]
Merge please? I don't think approve from somebody from Zephyr project is necessary. We use their system after all it matches their docs/tutorials so I think we are good to go.

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
Looks good to me. Lets wait for @marc-hb approval and will merge it.

**Review by marc-hb — COMMENTED** [review]
Sorry for not reviewing the shell script earlier, I think I found a small bug in it.

>  I don't think approve from somebody from Zephyr project is necessary. We use their system after all it matches their docs/tutorials so I think we are good to go.

Thank you for checking the Zephyr docs, really appreciated however docs are always lagging a bit behind and examples must always to be kept (too) generic. Also, this is a top-level directory meaning it would very painful to rename if needed, would rename many files with a near 100% chance of git conflict. So I have no objection but I would really like a nod from some Zephyr expert who's already done stuff like this. @andyross , @dcpleung could you help find someone? @mbolivar-nordic maybe?

Build-time configuration: what everyone loves to hate :-)

File: scripts/xtensa-build-zephyr.sh
Comment (aborisovich) [reply scripts/xtensa-build-zephyr.sh]: Fixed, thanks Marc! Bash scripts are not my strong side ;-)

**Comment by aborisovich** [comment]
Push update:

- fixed issues found by @marc-hb in the xtensa-build-zephyr.sh script.

**Comment by nashif** [comment]
How are those overlays going to be used if you decide to build using west or cmake directly and not use the wrapper scripts?

**Comment by andyross** [comment]
I guess I counta as a Zephyr voice.  Can I give a wishywashy answer here, skip all the specifics, and say I think that overlays look just fine?

There are a conceptually infinite number of ways to specify kconfig values for a Zephyr build.  By default you use a single "prj.conf" file.  You can use an alternate file instead via the CONF_FILE cmake variable.  Overlays provide a somewhat more flexible mechanism.

But remember it's also possible to do it in proper kconfig syntax (vs. the defconfig settings format used for prj.conf) in SOF's own kconfig tree by providing defaults for settings separate from their definitions.  That is, you could see that you're building on a 2-core TGL platform via an "if" and then add a bunch of defaults for Zephyr/SOF variables in that block, etc....  We do that in Zephyr at the soc and board level routinely, though not for apps (because apps don't have their own kconfig trees, but SOF does!).

Alternatively, I could see an architecture where SOF's build environment has a whole extra layer of configuration to auto-generate the prj.conf file from metaconfiguration too, and that doesn't sound so weird given the level of complexity.

I guess what I'm saying is that application configuration is a job for application code, and Zephyr's role is just to accomodate the scheme applications think works best.  Certainly if we're making this hard for you, we should view that as a bug and fix it.  But it's not our place to tell you how this should work. :)

**Comment by aborisovich** [comment]
So basically as @andyross indicated, its *our business* (SOFs) how we will generate the configuration Zephyr just gives interfaces to use. And I am not saying that overlays are final solution here especially that we have configs for non-zephyr platforms in other directory and other format. But overlays are the fastest/simplest way to do this for now by the book.

**Comment by marc-hb** [comment]
> How are those overlays going to be used if you decide to build using west or cmake directly and not use the wrapper scripts?

Does anyone ever invoke CMake directly? Every single time I "peeled the onion" and stepped down to invoking CMake directly it was to debug some CMake-related or other build issue.

I think the main reason every CMake project in the universe seems to have a CMake wrapper script (like for instance `west build`) is because CMake does not seem to support build configuration files,  at least not in any user friendly way. Developers need a single, simple and short command to build, I mean no one wants to type this hundreds of times per day:

```shell
cmake -DWEST_PYTHON=/usr/bin/python3 -B build-apl -S zephyr/samples/subsys/audio/sof -GNinja -DBOARD=intel_adsp_cavs15 -DOVERLAY=stuff -DTOOLCHAIN=blah -DFOO=bar...

ninja -C build-apl
# switch branch or something 
rm -rf build-apl

cmake -DWEST_PYTHON=/usr/bin/python3 -B build-apl -S zephyr/samples/subsys/audio/sof -GNinja -DBOARD=intel_adsp_cavs15 -DOVERLAY=stuff -DTOOLCHAIN=blah -DFOO=bar...
```

It does not take long for the (real! Try `west -v build`) CMake commands above to land in some kind of wrapper script, which is effectively a build configuration file in disguise. It's always faster to put together a quick and dirty wrapper script than to implement new configuration files and the parser that goes with them.

I see this PR as (among others) a preliminary step towards solving  part of this problem and removing some stuff from `xtensa-build-zephyr` (especially the toolchain configurations)

--- 

Now of course the other question is: why another layer of wrapper script (`xtensa-build-zephyr`) on top of `west` when  `west build` provides a wrapper on top of CMake _already_?
The other, current features of `xtensa-build-zephyr` are:

- Downloading zephyr. It's just a mere convenience for developers but it's critical for CI. If you can think of better ways to implement CI please share them in https://github.com/zephyrproject-rtos/zephyr/discussions/34713 and/or https://github.com/thesofproject/sof/issues/3517
- Building multiple configurations, similar to `twister --platform` (look, another wrapper script!). Again: convenience for developers but important for CI.
- Signing with rimage. Is there a single `west build-and-sign` command?
- Laying out `/lib/firmware/intel/` ready for deployment. Maybe this could/should be part of a custom `west flash` command?

So in the longer term maybe the last two can be transferred to west. In the shorter term, it's always faster to just dump your shell history into a script and the rest is history.

Not sure about the first two.

**Review by marc-hb — APPROVED** [review]

**Comment by lgirdwood** [comment]
@nashif @andyross I think we can refine and tune the build process as we converge more and identify/solve the opens. 

Our target should be we can invoke west on Windows/Linux and build, sign and optionally install our board + configuration + toolchain + key with a single simple (or maybe 2) west command(s). We will get there. Probably a topic for next call.

**Comment by marc-hb** [comment]
Thanks @nashif 

> Why do I need to download zephyr if I have it already downloaded and setup for my needs? 

You don't, it's an optional feature used mainly by CI.

> but I see this script being used for almost everything, like a one-stop shop, is never good idea. 

I don't think a "one-stop shop" is always a bad idea and this one is only 400 lines long and getting smaller. I'm more worried about features being implemented in the wrong place or duplicating some existing tool but if it comes to that then I'd rather have these design mistakes all contained in a single small script rather than scattered across multiple and even smaller scripts. With a single script at least CI and developers use and maintain the same script and build the same thing (the most common CI problem)

> Developers in zephyr do not use twister to build multiple images

If not using twister, how do developers build multiple platforms or multiple build configurations to make sure they didn't break any? It's notoriously easy to make "works4me" `CONFIG_` changes, hence `make randconfig` (which I think is missing from kconfiglib.py, I digress)

> Why can't you invoke signing in cmake directly?

It should be possible but `west sign` already exists and everyone tries to stay far away from CMake so it just happened like this. In fact SOF/XTOS is signed in CMake and good look making sense of the corresponding CMake code (I tried, several times. Check the SOF git log). There's also no potential build speedup here because you must re-sign for literally any source or configuration change.

> My original question is about how those overlays are going to be used if someone decides to use twister, west or cmake directly and not the wrapper script, why can't they be integrated in the lowest possible layer (cmake), so that every level of wrappers use the same information.

Thanks I understand better now. Very good question, exactly the type of questions I hoped people more familiar with Zephyr could help answer. How do other Zephyr projects solve this build configuration issue? Do they use Zephyr overlays too and if yes how do they make these overlays easy to pass to CMake without very long cmake command lines?

**Comment by nashif** [comment]
To further simplify and consolidate things I suggest moving the sample from zephyr into the sof tree and change zephyr/module.yaml to point where samples are. This way Zephyr's will be aware of it and CI will build it, see 

https://github.com/zephyrproject-rtos/mcuboot/blob/main/zephyr/module.yml

If you do it this way, you can put the overlays in the sample itself and will not have to tell cmake about them on the command line and put them in boards/ directory or in the CMakeFile.txt itself.

**Pull request merged**
